---
title: "Microbial_bioinformatic_analyses"
output: html_document
---

## loading R packages
```{r}
library(reshape2)
library(tidyverse)
library(ggplot2)
library(ggthemes)
library(vegan)
library(dplyr)
library(ggpubr) 
library(patchwork) 
library(ggsci)
library(pairwiseAdonis)
library(NetCoMi)
library(usedist)
library(microeco)
library(magrittr)
library(ggrepel)
library(dplyr)
library(tidytree)
library(ggtree)
library(ggpicrust2)
library(GGally)
library(ggprism)
library(patchwork)
library(ggh4x)
```


## Alpha diversity
```{r}
fuction_a=function(taxa,group){
Richness <- specnumber(taxa)
shannon <- diversity(taxa,"shannon")
chao1_estimates <- estimateR(taxa)
chao1=chao1_estimates[rownames(chao1_estimates)%in%c('S.chao1'),]

alpha <- cbind(group,Richness,shannon,chao1)
}

alpha=fuction_a(taxa,group)

table(alpha$Location)

color=c(pal_npg('nrc')(7),"#696969")
  
p <- ggplot(alpha, aes(x=Location, y=shannon,color=Location)) +
  geom_boxplot(aes(fill=Location),
               alpha=0.1,outlier.shape = NA)+
  geom_point(size = 2) + 
  scale_color_manual(values = c(color))+
  scale_fill_manual(values = c(color))+
  theme_bw()+
  theme(panel.grid = element_blank())+ stat_compare_means(method="kruskal.test")+ #or wilcox.test
  theme(axis.line.x = element_blank(), 
        axis.text.x  = element_text(size=10),
       axis.title.x = element_blank(),
       axis.title.y = element_text(size=12, face="bold")) +
labs( y = "Shannon index")

```


## Beta diversity
```{r}
bray_analysis=as.matrix(vegdist(taxa, "bray"))
set.seed(625)
ado<-adonis(bray_analysis ~ group, data = analysis)
ado
summary(ado)
pval=ado$aov.tab$`Pr(>F)`[1]
R2=round(ado$aov.tab$R2[1],2)

pcoa = cmdscale(bray_analysis, k=3, eig=T) # k is dimension, 3 is recommended; eig is eigenvalues
points = as.data.frame(pcoa$points) # get coordinate string, format to dataframme
eig = pcoa$eig
points = cbind(points, analysis)
colnames(points)[1:3] = c("x", "y", "z") 

main_theme = theme(panel.background=element_blank(),
                   panel.grid=element_blank(),
                   axis.line.x=element_line(size=.5, colour="black"),
                   axis.line.y=element_line(size=.5, colour="black"),
                   axis.ticks=element_line(color="black"),
                   axis.text=element_text(color="black", size=10),
                   legend.position="right",
                   legend.background=element_blank(),
                   legend.key=element_blank(),
                   legend.text= element_text(size=10),
                   text=element_text(family="sans", size=10),
                   plot.title = element_text(hjust = 0.5,size = 14,face="bold"))


p = ggplot(points, aes(x=x, y=y, color=group)) + 
  geom_point(alpha=.7, size=2) +
  labs(x=paste("PCoA 1 (", format(100 * eig[1] / sum(eig), digits=4), "%)", sep=""),
       y=paste("PCoA 2 (", format(100 * eig[2] / sum(eig), digits=4), "%)", sep=""))+
  stat_ellipse(level=0.68)+
  scale_color_manual(values = c(color))+
  main_theme+ggtitle("PCoA plot")+
  annotate("text",x = -Inf, y = Inf, label = paste(paste0("R2=",R2),paste0("P =", pval)), hjust = -.2, vjust = 1)

```


## Top bacteria
```{r}
function_prop=function(input,group,name){
taxa=as.data.frame(t(input))
taxa=taxa[rownames(group),]

df_TSTM=taxa[rownames(group[group$Location%in%c('TSTM'),]),]
df_NOZ=taxa[rownames(group[group$Location%in%c('NOZ'),]),]
df_HNZ=taxa[rownames(group[group$Location%in%c('HNZ'),]),]
df_RZ=taxa[rownames(group[group$Location%in%c('RZ'),]),]
df_NT=taxa[rownames(group[group$Location%in%c('NT'),]),]
df_PB=taxa[rownames(group[group$Location%in%c('PB'),]),]

df_TSTM<- df_TSTM[,colSums(df_TSTM)>0]
df_NOZ<- df_NOZ[,colSums(df_NOZ)>0]
df_HNZ<- df_HNZ[,colSums(df_HNZ)>0]
df_RZ<- df_RZ[,colSums(df_RZ)>0]
df_NT<- df_NT[,colSums(df_NT)>0]
df_PB<- df_PB[,colSums(df_PB)>0]

df_TSTM_rb <- df_TSTM/rowSums(df_TSTM)
df_NOZ_rb <- df_NOZ/rowSums(df_NOZ)
df_HNZ_rb <- df_HNZ/rowSums(df_HNZ)
df_RZ_rb <- df_RZ/rowSums(df_RZ)
df_NT_rb <- df_NT/rowSums(df_NT)
df_PB_rb <- df_PB/rowSums(df_PB)

function_top <- function(in_matrix, identifier, top_no){
  colsum <- colSums(in_matrix)
  colsum <- colsum[order(colsum, decreasing = TRUE)]
  top10 <- head(names(colsum), top_no)
  df_top10 <- in_matrix[,top10]
  df_others <- rowSums(in_matrix[,! names(in_matrix) %in% top10])
  df_top10$Others <- df_others
  df_top10$id <- rownames(df_top10)
  df_top10_melt <- reshape2::melt(df_top10, id="id")
  df_top10_melt[,"identifier"] <- identifier
  return(df_top10_melt)
}

df_TSTM_rb_top <- function_top(df_TSTM_rb, "TSTM", 10)
df_NOZ_rb_top <- function_top(df_NOZ_rb, "NOZ", 10)
df_HNZ_rb_top <- function_top(df_HNZ_rb, "HNZ", 10)
df_RZ_rb_top <- function_top(df_RZ_rb, "RZ", 10)
df_NT_rb_top <- function_top(df_NT_rb, "NT", 10)
df_PB_rb_top <- function_top(df_PB_rb, "PB", 10)

df_combine <- rbind(df_TSTM_rb_top, df_NOZ_rb_top,
                    df_HNZ_rb_top,df_RZ_rb_top,df_NT_rb_top,df_PB_rb_top )
df_combine$identifier <- factor(df_combine$identifier, levels = c("TSTM", "NOZ", "HNZ", "RZ","NT","PB"))
df_order <- doBy::summaryBy(value~variable, data=df_combine[,c("variable", "value")], FUN=sum)
df_order <- df_order[order(df_order$value.sum, decreasing = TRUE),]
df_order_moved_rows <- subset(df_order,variable=="Others")
df_order_remaining_rows <- subset(df_order,variable!="Others")
df_order_final <- rbind(df_order_remaining_rows, df_order_moved_rows)
df_combine$variable <- factor(df_combine$variable, levels = df_order_final$variable)

p_top <- ggplot(df_combine, aes(x=id,  y=value)) + 
  geom_area(aes(fill = variable, group = variable))+
  facet_wrap(vars(identifier), nrow = 6) +
  scale_fill_d3("category20")+
  theme_bw()+
  xlab("")+
  ylab(paste0("Relative abundance of ",name))+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

p_top 
return(p_top)
}

p_top_phy=function_prop(phy,group,"Phylum")
p_top_gen=function_prop(gen,group,"Genus")

```



## Aitchison dissimilarities
```{r}
function_distance=function(meta,Patient){
taxa=t(spe[,rownames(meta)])

dissMat = taxa %>% 
  NetCoMi::netConstruct(., 
                        measure = "euclidean",
                        zeroMethod = "none",
                        normMethod = "mclr", 
                        cores = 24,
                        sparsMethod = "none", 
                        seed = 123456)


betadisper.primary= vegan::betadisper(as.dist(dissMat$dissMat1), meta$Location, bias.adjust = F) %>% 
  with(., dist(centroids))

write.csv(as.matrix(betadisper.primary),file=paste0(paste0(Patient,'_Aitchison.csv')))

}

meta=group[group$Patient%in%c('Patient1'),]
function_distance(meta,'Patient1')

meta=group[group$Patient%in%c('Patient2'),]
function_distance(meta,'Patient2')

meta=group[group$Patient%in%c('Patient3'),]
function_distance(meta,'Patient3')

meta=group[group$Patient%in%c('Patient4'),]
function_distance(meta,'Patient4')

meta=group[group$Patient%in%c('Patient5'),]
function_distance(meta,'Patient5')

meta=group[group$Patient%in%c('Patient6'),]
function_distance(meta,'Patient6')

meta=group[group$Patient%in%c('Patient7'),]
function_distance(meta,'Patient7')

meta=group[group$Patient%in%c('Patient8'),]
function_distance(meta,'Patient8')

meta=group[group$Patient%in%c('Patient9'),]
function_distance(meta,'Patient9')

meta=group[group$Patient%in%c('Patient10'),]
function_distance(meta,'Patient10')

meta=group[group$Patient%in%c('Patient11'),]
function_distance(meta,'Patient11')

meta=group[group$Patient%in%c('Patient12'),]
function_distance(meta,'Patient12')

meta=group[group$Patient%in%c('Patient13'),]
function_distance(meta,'Patient13')

meta=group[group$Patient%in%c('Patient14'),]
function_distance(meta,'Patient14')

meta=group[group$Patient%in%c('Patient15'),]
function_distance(meta,'Patient15')

meta=group[group$Patient%in%c('Patient16'),]
function_distance(meta,'Patient16')

meta=group[group$Patient%in%c('Patient17'),]
function_distance(meta,'Patient17')


Ait_long <- melt(Ait, id.vars = "Location", variable.name = "Sample", value.name = "Value")
Ait_long$Location=factor(Ait_long$Location,levels =c("TSTM&RZ","TSTM&NT","TSTM&PB",
                                           "NOZ&RZ","NOZ&NT","NOZ&PB",
                                           "HNZ&RZ","HNZ&NT","HNZ&PB"))
my_comparisons <- list(c("TSTM&RZ","TSTM&NT"),c("TSTM&NT","TSTM&PB"),c("TSTM&RZ","TSTM&PB"),
                       c("NOZ&RZ","NOZ&NT"),c("NOZ&NT","NOZ&PB"),c("NOZ&RZ","NOZ&PB"),
                      c("HNZ&RZ","HNZ&NT"),c("HNZ&NT","HNZ&PB"),c("HNZ&RZ","HNZ&PB") )

p <-ggplot(Ait_long, aes(x=Location, y=Value,color=Location)) +
  geom_boxplot(aes(fill=Location),
               alpha=0.4,outlier.shape = NA)+ 
  geom_jitter()+
  theme_bw()+
  theme(panel.grid = element_blank())+   stat_compare_means(comparisons = my_comparisons,
                     method = "wilcox.test")+
  theme(axis.line.x = element_blank(), 
        axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
       axis.title.x = element_blank(),
       axis.title.y = element_text(size=12, face="bold")) +
labs( y = "Aitchinson similarity")

p


color_gradient <- colorRampPalette(c('#EFF2BA','#80BFE8','#9B74A4'))(50)
primary.heatmap = ggcorrplot::ggcorrplot(as.matrix(Ait_mean), type = "full",lab=T) +
    scale_fill_gradientn(colors = color_gradient, 
                         name = "Aitchinson\nDissimilarity") +
    theme_few(base_size = 12) +
    theme(legend.position = c(1.2, 0.25)) +
    ggpubr::rotate_x_text(angle = 45) +
    xlab("") +
    ylab("")

```


## The linear discriminant analysis (LDA) effect size (LEfSe)
```{r}
ASV=gen
TAX=as.data.frame(rownames(gen))
ASV <- apply(ASV,2,function(x) x/sum(x))

df <- microtable$new(sample_table = GROUP,
                     otu_table = ASV,
                     tax_table = TAX,
                     auto_tidy = FALSE)

lefse <- trans_diff$new(dataset = df,
                        method = "lefse",
                        group = "Group",
                        alpha = 0.05)

result=lefse$res_diff

color=pal_locuszoom("default", alpha = 0.8)(7)
Tumorcolor=color[2]
noncolor=color[4]

p=lefse$plot_diff_bar(threshold = 2,
                    width = 0.6,
                    alpha = 0.8,
                    group_order = c('Tumor','Non-tumor')) +
  scale_color_manual(values = c(Tumorcolor,noncolor)) +
  scale_fill_manual(values =  c(Tumorcolor,noncolor)) +
  theme(legend.key = element_blank(),
        legend.title = element_blank(),
        legend.background = element_rect(fill = 'white'),
        legend.justification = c(1, 0),
        legend.position = c(1, 0.001))

p

```


## Phylogenetic Investigation of Communities by Reconstruction of Unobserved States 2 (PICRUSt2) visualization
```{r}
data_ko <- read.table("KO_metagenome_out/pred_metagenome_unstrat.tsv.gz", header=T, sep="\t")
ko_input=data_ko[,c("function.",rownames(group))]
kegg_abundance<-ko2kegg_abundance(data=ko_input)
daa_results_df<-pathway_daa(kegg_abundance,metadata=group_input,
                            group="Group",daa_method="LinDA")
daa_annotated_results_df<-pathway_annotation(pathway="KO",
                                             daa_results_df=daa_results_df,ko_to_kegg=TRUE)
p<-pathway_errorbar(abundance=kegg_abundance,
                    daa_results_df=daa_annotated_results_df,
                    Group=group_input$Group,
                    ko_to_kegg=TRUE,
                    p_values_threshold=0.05,
                    order="pathway_class",
                    select=NULL,
                    p_value_bar=TRUE,
                    colors=NULL,
                    x_lab="pathway_name")

p

```
